timeout_handler.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package simplex
+
+import (
+	"container/heap"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type TimeoutTask struct {
+	NodeID   NodeID
+	TaskID   string
+	Start    uint64
+	End      uint64
+	Task     func()
+	Deadline time.Time
+
+	index int // for heap to work more efficiently
+}
+
+type TimeoutHandler struct {
+	lock sync.Mutex
+
+	ticks chan time.Time
+	close chan struct{}
+	// nodeids -> range -> task
+	tasks map[string]map[string]*TimeoutTask
+	heap  TaskHeap
+	now   time.Time
+
+	log Logger
+}
+
+// NewTimeoutHandler returns a TimeoutHandler and starts a new goroutine that
+// listens for ticks and executes TimeoutTasks.
+func NewTimeoutHandler(log Logger, startTime time.Time, nodes []NodeID) *TimeoutHandler {
+	tasks := make(map[string]map[string]*TimeoutTask)
+	for _, node := range nodes {
+		tasks[string(node)] = make(map[string]*TimeoutTask)
+	}
+
+	t := &TimeoutHandler{
+		now:   startTime,
+		tasks: tasks,
+		ticks: make(chan time.Time, 1),
+		close: make(chan struct{}),
+		log:   log,
+	}
+
+	go t.run()
+
+	return t
+}
+
+func (t *TimeoutHandler) GetTime() time.Time {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.now
+}
+
+func (t *TimeoutHandler) run() {
+	for t.shouldRun() {
+		select {
+		case now := <-t.ticks:
+			t.lock.Lock()
+			t.now = now
+			t.lock.Unlock()
+
+			t.maybeRunTasks()
+		case <-t.close:
+			return
+		}
+	}
+}
+
+func (t *TimeoutHandler) maybeRunTasks() {
+	// go through the heap executing relevant tasks
+	for {
+		t.lock.Lock()
+		if t.heap.Len() == 0 {
+			t.lock.Unlock()
+			break
+		}
+
+		next := t.heap[0]
+		if next.Deadline.After(t.now) {
+			t.lock.Unlock()
+			break
+		}
+
+		heap.Pop(&t.heap)
+		delete(t.tasks[string(next.NodeID)], next.TaskID)
+		t.lock.Unlock()
+		t.log.Debug("Executing timeout task", zap.String("taskid", next.TaskID))
+		next.Task()
+	}
+}
+
+func (t *TimeoutHandler) shouldRun() bool {
+	select {
+	case <-t.close:
+		return false
+	default:
+		return true
+	}
+}
+
+func (t *TimeoutHandler) Tick(now time.Time) {
+	select {
+	case t.ticks <- now:
+	default:
+		t.log.Debug("Dropping tick in timeouthandler")
+	}
+}
+
+func (t *TimeoutHandler) AddTask(task *TimeoutTask) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, ok := t.tasks[string(task.NodeID)]; !ok {
+		t.log.Info("Attempting to add a task for an unknown node", zap.Stringer("node", task.NodeID))
+	}
+
+	// adds a task to the heap and the tasks map
+	if _, ok := t.tasks[string(task.NodeID)][task.TaskID]; ok {
+		t.log.Debug("Trying to add an already included task", zap.Stringer("node", task.NodeID), zap.String("Task ID", task.TaskID))
+		return
+	}
+
+	t.tasks[string(task.NodeID)][task.TaskID] = task
+	t.log.Debug("Adding timeout task", zap.Stringer("node", task.NodeID), zap.String("taskid", task.TaskID))
```

Comment:
nit, but when i run the test i get:

```
[04-24|00:10:19.239] DEBUG Simplex/timeout_handler.go:137 Adding timeout task {"test": "TestReplicationRequestTimeoutMultiple", "node": 108, "node": "0100000000000000", "taskid": "0_3"}
```

Can we change the `node` here to something else like `from` ? This is confusing...